File: plugins/plugin-tlon/rust/src/actions/send_message.rs
//! Send message action.

use async_trait::async_trait;
use serde_json::Value;

use super::{ActionContext, TlonAction};
use crate::error::Result;

/// Action that builds a JSON payload for sending a Tlon message.
pub struct SendMessageAction;

#[async_trait]
impl TlonAction for SendMessageAction {
    fn name(&self) -> &'static str {
        "SEND_TLON_MESSAGE"
    }

    fn description(&self) -> &'static str {
        "Send a message via Tlon/Urbit to a ship or channel"
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        let source = context.message.get("source").and_then(|v| v.as_str());
        Ok(source == Some("tlon") || source == Some("urbit"))
    }

    async fn execute(&self, context: &ActionContext) -> Result<Value> {
        let response_text = context
            .state
            .get("response")
            .and_then(|v| v.get("text"))
            .and_then(|v| v.as_str())
            .unwrap_or("");

        let target = if let Some(ref nest) = context.channel_nest {
            serde_json::json!({ "channel_nest": nest })
        } else if let Some(ref ship) = context.ship {
            serde_json::json!({ "ship": ship })
        } else {
            serde_json::json!({})
        };

        Ok(serde_json::json!({
            "action": self.name(),
            "target": target,
            "text": response_text,
            "reply_to_id": context.reply_to_id
        }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_validate_tlon_source() {
        let action = SendMessageAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "tlon",
                "text": "Hello"
            }),
            ship: Some("sampel-palnet".to_string()),
            channel_nest: None,
            reply_to_id: None,
            state: serde_json::json!({}),
        };

        assert!(action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_urbit_source() {
        let action = SendMessageAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "urbit",
                "text": "Hello"
            }),
            ship: None,
            channel_nest: Some("chat/~host/channel".to_string()),
            reply_to_id: None,
            state: serde_json::json!({}),
        };

        assert!(action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_other_source() {
        let action = SendMessageAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "telegram",
                "text": "Hello"
            }),
            ship: None,
            channel_nest: None,
            reply_to_id: None,
            state: serde_json::json!({}),
        };

        assert!(!action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_execute_dm() {
        let action = SendMessageAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "tlon"
            }),
            ship: Some("sampel-palnet".to_string()),
            channel_nest: None,
            reply_to_id: None,
            state: serde_json::json!({
                "response": { "text": "Hello from bot!" }
            }),
        };

        let result = action.execute(&context).await.unwrap();
        assert_eq!(result["action"], "SEND_TLON_MESSAGE");
        assert_eq!(result["target"]["ship"], "sampel-palnet");
        assert_eq!(result["text"], "Hello from bot!");
    }
}

File: plugins/plugin-tlon/typescript/src/actions/sendMessage.ts
import type {
  Action,
  ActionExample,
  ActionResult,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { TLON_SERVICE_NAME } from "../constants";
import type { TlonService } from "../service";
import type { TlonContent } from "../types";

export const SEND_MESSAGE_ACTION = "SEND_TLON_MESSAGE";

/**
 * Action for sending messages via Tlon/Urbit
 */
export const sendMessageAction: Action = {
  name: SEND_MESSAGE_ACTION,
  similes: [
    "TLON_SEND_MESSAGE",
    "TLON_REPLY",
    "TLON_MESSAGE",
    "SEND_TLON",
    "REPLY_TLON",
    "URBIT_SEND_MESSAGE",
    "URBIT_MESSAGE",
    "SEND_URBIT",
  ],
  description: "Send a message via Tlon/Urbit to a ship or channel",

  validate: async (
    _runtime: IAgentRuntime,
    message: Memory,
  ): Promise<boolean> => {
    const source = message.content?.source;
    return source === "tlon" || source === "urbit";
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const tlonService = (await runtime.getService(TLON_SERVICE_NAME)) as
      | TlonService
      | undefined;

    if (!tlonService) {
      if (callback) {
        await callback({
          text: "Tlon service not available",
        });
      }
      return { success: false, error: "Tlon service not initialized" };
    }

    const responseText = state?.values?.response?.toString() || "";
    const content = message.content as TlonContent;
    const channelId =
      content?.channelNest || (content as { chatId?: string })?.chatId;
    const ship = content?.ship;
    const replyToId = content?.replyToId;

    // Determine if this is a DM or channel message
    const isDm = ship && !channelId?.includes("/");

    try {
      if (isDm && ship) {
        await tlonService.sendDirectMessage(ship, responseText);
      } else if (channelId) {
        await tlonService.sendChannelMessage(
          channelId,
          responseText,
          replyToId,
        );
      } else {
        if (callback) {
          await callback({
            text: "No target ship or channel specified",
          });
        }
        return { success: false, error: "Missing target ship or channel" };
      }

      if (callback) {
        await callback({
          text: responseText,
          action: SEND_MESSAGE_ACTION,
        });
      }

      return {
        success: true,
        data: {
          action: SEND_MESSAGE_ACTION,
          channelId: channelId || ship,
          text: responseText,
          replyToId,
        },
      };
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      if (callback) {
        await callback({
          text: `Failed to send message: ${errorMessage}`,
        });
      }
      return { success: false, error: errorMessage };
    }
  },

  examples: [
    [
      {
        name: "{{name1}}",
        content: {
          text: "Send a message to this Tlon chat",
        },
      },
      {
        name: "{{agentName}}",
        content: {
          text: "I'll send a message via Tlon now.",
          actions: [SEND_MESSAGE_ACTION],
        },
      },
    ],
    [
      {
        name: "{{name1}}",
        content: {
          text: "Reply to this message on Urbit",
        },
      },
      {
        name: "{{agentName}}",
        content: {
          text: "I'll reply to your message on Urbit.",
          actions: [SEND_MESSAGE_ACTION],
        },
      },
    ],
  ] as ActionExample[][],
};
